docs/Dialects/Verif.md

```diff
@@ -4,6 +4,334 @@ This dialect provides a collection of operations to express various verification
+
+### Carry-Save Compressor Example
+
+Consider the following slightly more involved example of a compression stage you would find in a carry-save adder.
+This module takes 3 input values and produces 2 output values that sum up to the same value as its inputs.
+However, the contract in the module does not specify which _exact_ output values the module produces.
+Instead, it uses two symbolic values to express that the module can produce _any_ output values that sum up to the inputs.
+In a sense, how exactly the compressor combines 3 values into 2 is left as an implementation detail that you can’t know about, but the guarantee you can work with is that the sum will be correct.
+This is different from the previous example, where the contract produced an exact replacement value for the module output.
+Also, note how this uses an `assume` to constrain the sum of the symbolic values instead of an `ensure` or `require`.
+
+```mlir
+// A module that takes 3 input values and produces 2 output values that sum up
+// to the same value as the inputs. Instead of just using add it uses a
+// bit-parallel full adder that takes each 3-tuple of bits in the 3 inputs, runs
+// them through a full adder, and treats the resulting sum and carry as the 2
+// corresponding bits for its 2 output values.
+hw.module @CarrySaveCompress3to2(
+  in %a0: i42, in %a1: i42, in %a2: i42,
+  out z0: i42, out z1: i42
+) {
+  %c1_i42 = hw.constant 1 : i42
+  %0 = comb.xor %a0, %a1, %a2 : i42  // sum bits of FA (a0^a1^a2)
+  %1 = comb.and %a0, %a1 : i42
+  %2 = comb.or %a0, %a1 : i42
+  %3 = comb.and %2, %a2 : i42
+  %4 = comb.or %1, %3 : i42          // carry bits of FA (a0&a1 | a2&(a0|a1))
+  %5 = comb.shl %4, %c1_i42 : i42    // %5 = carry << 1
+  // At this point, %0+%5 is the same as %a0+%a1+%a2, but without creating a
+  // long ripple-carry chain.
+
+  // Contract to check that we output _some_ two numbers that sum up to the same
+  // value as the sum of the three inputs. We don't say which exact numbers.
+  %z0, %z1 = verif.contract %0, %5 {
+    // The contract promises that its outputs will sum up to the same value as
+    // the sum of the module inputs.
+    %inputSum = comb.add %a0, %a1, %a2 : i42
+    %outputSum = comb.add %z0, %z1 : i42
+    verif.ensure_equal %inputSum, %outputSum : i42
+
+    // We don't want this contract to give guarantees about what the exact
+    // values of its outputs are going to be. Instead, we only want to guarantee
+    // that they sum up to the right number. To express this, we pick two
+    // symbolic values and constrain them to sum up to that number, and yield
+    // those from the contract. This says "you'll get any two numbers that sum
+    // up to the sum of the inputs".
+    %any0 = verif.symbolic_value : i42
+    %any1 = verif.symbolic_value : i42
+    %anySum = comb.add %any0, %any1 : i42
+    verif.assume_equal %anySum, %inputSum : i42
```

Comment:
Hmm, that's a very interesting thought. Maybe these two things are separate. The idea behind the contracts was to automate the process of assert-assume swapping, to ensure that you're actually only assuming what you've asserted before. I wonder if the substitutions are an orthogonal piece of the puzzle :thinking: How would you split the above up into a contract and a potentially new op that encodes substitution?

Reply:
I think the contract op could almost stay as is. Just remove the operands and results.
For the op that provides the alternative, I'm not exactly sure. If we'd specify contracts on the module level, I'd say, just put it in another module and link them with some `verif.alternative` op. But it'd also work to use the same structure as the contract has now, just without the `ensure` and `require` ops inside.

I'm generally also wondering if the contract doesn't have to specify more explicitly what part of the module it applies to. E.g., if you have a memory write operation. It wouldn't show up in a DFG when you just consider the fan-in of some SSA value. Similarly, for things like `llhd.drv` or anything that uses connect or alloca semantics in some way. I'm wondering how bad it would actually be to force a contract to apply to exactly one whole module? It doesn't necessarily have to be a module at the Chisel level, just at the HW level. The compiler can just carve out the part of a module the contract applies to and factor it out into its own private submodule.

```mlir
// Applies to the module implementation, and also every `verif.alternative` by
// inlining the alternative body into the module the way you currently inline contracts.
// What if there are multiple contracts and/or multiple alternatives?
// I think forcing them to operate on a per-module basis would simplify things greatly.
// You'd have 0 or 1 contract per module and any number of alternative
// implementations per module that cannot overlap.
verif.contract {
  %inputSum = comb.add %a0, %a1, %a2 : i42
  %outputSum = comb.add %z0, %z1 : i42
  verif.ensure_equal %inputSum, %outputSum : i42
}
%z0, %z1 = verif.alternative %0, %5 {
  %inputSum = comb.add %a0, %a1, %a2 : i42
  %any0 = verif.symbolic_value : i42
  %any1 = verif.symbolic_value : i42
  %anySum = comb.add %any0, %any1 : i42
  verif.assume_equal %anySum, %inputSum : i42
  verif.yield %any0, %any1 : i42, i42
}
```

Or a module-based design:
```mlir
module {
  // Written by the verification engineer or auto-generated from some spec. Makes it easier to put it in a different input file than the implementation by just having some `verif.contract.extern`. Design engineers can then implement several versions of designs that implement the same contract.
  verif.contract @contract(in %a0: i42, in %a1: i42, in %a2: i42, out %z0: i42, out %z1: i42) {
    %inputSum = comb.add %a0, %a1, %a2 : i42
    %outputSum = comb.add %z0, %z1 : i42
    verif.ensure_equal %inputSum, %outputSum : i42
  }

  // Verification flows look up the contract and set up a 'verif.formal' like this:
  // verif.formal {
  //   %a0 = verif.symbolic_value : i42
  //   %a1 = verif.symbolic_value : i42
  //   %a2 = verif.symbolic_value : i42
  //   %z0, %z1 = hw.instance @CarrySaveCompress3to2(%a0: i42, %a1: i42, %a2: i42) -> (z0: i42, z1: i42)
  //   %inputSum = comb.add %a0, %a1, %a2 : i42
  //   %outputSum = comb.add %z0, %z1 : i42
  //   verif.assert_equal %inputSum, %outputSum : i42
  // }
  hw.module @CarrySaveCompress3to2(in %a0: i42, in %a1: i42, in %a2: i42, out z0: i42, out z1: i42) implements @contract {
    %c1_i42 = hw.constant 1 : i42
    %0 = comb.xor %a0, %a1, %a2 : i42  // sum bits of FA (a0^a1^a2)
    %1 = comb.and %a0, %a1 : i42
    %2 = comb.or %a0, %a1 : i42
    %3 = comb.and %2, %a2 : i42
    %4 = comb.or %1, %3 : i42          // carry bits of FA (a0&a1 | a2&(a0|a1))
    %5 = comb.shl %4, %c1_i42 : i42    // %5 = carry << 1
     hw.output %0, %5 : i42, i42
  }

  // same for this alternative implementation
  hw.module @alt1(in %a0: i42, in %a1: i42, in %a2: i42, out z0: i42, out z1: i42) implements @contract {
    %inputSum = comb.add %a0, %a1, %a2 : i42
    %any0 = verif.symbolic_value : i42
    %any1 = verif.symbolic_value : i42
    %anySum = comb.add %any0, %any1 : i42
    verif.assume_equal %anySum, %inputSum : i42
    verif.yield %any0, %any1 : i42, i42
  }

  hw.module @top(in %a0: i42, in %a1: i42, in %a2: i42, out z0: i42, out z1: i42) {
    // Regular ExportVerilog/Synthesis flows just treat it as a regular instance and ignore the 'implements' attribute
    // Formal verification flows look up the module, see the 'implements' attribute, look up the 'contract', and inline it here by inserting symbolic values for all instance outputs and replacing 'ensure' -> 'assume'; 'require' -> 'assert'
    // It's the contract's job to constrain these symbolic values enough that the contract this @top module is implementing can be verified.
    %z0, %z1 = hw.instance @CarrySaveCompress3to2(%a0: i42, %a1: i42, %a2: i42) -> (z0: i42, z1: i42)
    // This results in:
    // %inputSum = comb.add %a0, %a1, %a2 : i42
    // %z0 = verif.symbolic_value : i42
    // %z1 = verif.symbolic_value : i42
    // %outputSum = comb.add %z0, %z1 : i42
    // verif.assume_equal %inputSum, %outputSum : i42
    hw.output %z0, %z1 : i42, i42
  }
  // Or, use the alternative implementation instead (need to iterate over all modules and check if it implements the same contract; maybe we can add another hint to indicate for what the alternative is intended because it could also be a golden model, symbolic implementation, simulation optimized implementation, etc.):
  hw.module @top(in %a0: i42, in %a1: i42, in %a2: i42, out z0: i42, out z1: i42) {
    %z0, %z1 = hw.instance @alt1(%a0: i42, %a1: i42, %a2: i42) -> (z0: i42, z1: i42)
    hw.output %z0, %z1 : i42, i42
  }
}
```